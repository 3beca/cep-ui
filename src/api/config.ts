export const BASE_URL = process.env.REACT_APP_BASE_URL || 'https://cep.tribeca.ovh';
export const EVENT_TYPE_URL = '/admin/event-types';