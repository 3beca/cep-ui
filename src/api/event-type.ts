import {
    fetchData,
    RequestInfo,
    ResponseData,
    ErrorAPI,
    ResponseEmptyData
} from './fetch-data';
import {
    BASE_URL,
    EVENT_TYPE_URL
} from './config';
export * from './fetch-data';

export type EventType = {
    name: string;
    id: string;
    url: string;
    createdAt: string;
    updatedAt: string;
};
export type EventTypeList = {
    results: EventType[];
    next?: string | undefined;
    prev?: string | undefined;
};
export const getEventTypeList = (config: RequestInfo) => async (page: number = 1, size: number = 10): Promise<ResponseData<EventTypeList>|ErrorAPI> => {
    const url = `${config.url}${EVENT_TYPE_URL}/?page=${page}&pageSize=${size}`;
    return await fetchData<undefined, EventTypeList>({...config, url, method: 'GET'});
};
export const deleteEventType = (config: RequestInfo) => async (eventTypeIds: string|string[]): Promise<(ResponseEmptyData|ErrorAPI)[]> => {
    const eventTypesArray = (Array.isArray(eventTypeIds)) ? eventTypeIds : [eventTypeIds];
    if(!eventTypeIds || eventTypesArray.length < 1) {
        return [{
            status: 500,
            error: 'Missing EventType id',
            message: 'eventTypeIds is an invalid EventTypeId value or array'
        }];
    }
    const requests = [];
    for (const eventTypeId of eventTypesArray) {
        const url = `${config.url}${EVENT_TYPE_URL}/${eventTypeId}`;
        requests.push(fetchData<undefined, undefined>({...config, url, method: 'DELETE'}));
    }
    return Promise.all(requests);
};
export type EventTypeApi = {
    getEventTypeList(page?: number, size?: number): Promise<ResponseData<EventTypeList>|ErrorAPI>;
    deleteEventType(eventTypeIds: string|string[]): Promise<(ResponseEmptyData|ErrorAPI)[]>;
};
export const EventTypeApi = (): EventTypeApi => {
    const config: RequestInfo = {
        url: BASE_URL,
        method: 'GET',
        headers: {}
    };
    return {
        getEventTypeList: getEventTypeList(config),
        deleteEventType: deleteEventType(config)
    };
};
