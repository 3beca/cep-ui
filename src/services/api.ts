import {
    fetchApi,
    APIRequestInfo,
    APIResponseData,
    APIError,
    isAPIError
} from '../utils/fetch-api';

export type ServiceList<T> = {
    results: T[];
    next?: string | undefined;
    prev?: string | undefined;
};
export type ServiceError = {
    statusCode: number,
    error: string,
    message: string
};
export type Entity = {
    id: string;
};
export type WithDate = {
    createdAt: string;
    updatedAt: string;
};
export type ServiceDeleted = {
    state: 'DELETED'|'REJECTED';
    error?: ServiceError;
} & Entity;
export type EventType = {
    name: string;
    url: string;
} & Entity & WithDate;
export type EventTypeList = ServiceList<EventType>;
export type EventTypeError = ServiceError;
export type EventTypeDeleted = ServiceDeleted;

export type Target = {
    name: string;
    url: string;
} & Entity & WithDate;
export type TargetList = ServiceList<Target>;
export type TargetError = ServiceError;
export type TargetDeleted = ServiceDeleted;

export type RuleTypes = 'sliding'|'hopping'|'tumbling'|'none';
export type Rule = {
    name: string;
    type: RuleTypes;
    targetId: string;
    eventTypeId: string;
    skipOnConsecutivesMatches?: boolean;
    filters: any;
} & Entity & WithDate;
export type RuleList = ServiceList<Rule>;
export type RuleError = ServiceError;
export type RuleDeleted = ServiceDeleted;

export type EventLog = {
    eventTypeId: string;
    eventTypeName: string;
    payload: {[key: string]: any;};
    requestId: string;
    createdAt: string;
    rules: {
        id: string;
        name: string;
        targetId: string;
    }[];
    targets: {
        id: string;
        name: string;
        response: {
            statusCode: number;
            body: {[key: string]: any;};
        };
    }[];
} & Entity;

export const getListRequest = (baseURL: string, config: APIRequestInfo) => async <T>(path: string, page: number = 1, size: number = 10, filter: string = ''): Promise<APIResponseData<ServiceList<T>>|APIError<ServiceError>> => {
    const url = `${baseURL}${path}/?page=${page}&pageSize=${size}${filter ? `&search=${filter}` : ''}`;
    return fetchApi<undefined, ServiceList<T>, ServiceError>(url, {...config, method: 'GET'});
};
export const deleteRequest = (baseURL: string, config: APIRequestInfo) => async (path: string, eventTypeIds: string|string[]): Promise<(APIResponseData<ServiceDeleted[]>|APIError<ServiceError>)> => {
    const eventTypesArray = ((Array.isArray(eventTypeIds)) ? eventTypeIds : [eventTypeIds]).filter(id => !!id);

    if(!eventTypeIds || eventTypesArray.length < 1) {
        return {
            errorCode: 500,
            errorMessage: '',
            error: {
                statusCode: 500,
                error: 'Missing id',
                message: 'id is an invalid id value or array'
            }
        };
    }
    const requests = [];
    for (const eventTypeId of eventTypesArray) {
        const url = `${baseURL}${path}/${eventTypeId}`;
        requests.push(fetchApi<undefined, undefined, ServiceError>(url, {...config, method: 'DELETE'}));
    }
    const responses = await Promise.all(requests);
    const data: ServiceDeleted[] = responses.map((response, idx) => {
        const deletedResponse: ServiceDeleted = {
            id: eventTypesArray[idx],
            state: 'DELETED'
        };
        if(isAPIError(response)) {
            deletedResponse.state = 'REJECTED';
            deletedResponse.error = response.error;
        }
        return deletedResponse;
    });
    return {
        status: 200,
        data
    };
};
export type Api = {
    getListRequest<T>(path: string, page?: number, size?: number, filter?: string): Promise<APIResponseData<ServiceList<T>>|APIError<ServiceError>>;
    deleteRequest(path: string, ids: string|string[]): Promise<(APIResponseData<ServiceDeleted[]>|APIError<ServiceError>)>;
};
export const buildApiService = (server: string): Api => {
    const config: APIRequestInfo = {
        method: 'GET',
        headers: {}
    };
    return {
        getListRequest: getListRequest(server, config),
        deleteRequest: deleteRequest(server, config)
    };
};
