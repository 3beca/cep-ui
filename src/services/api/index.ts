import {
    fetchApi,
    APIRequestInfo,
    APIResponseData,
    APIError,
    isAPIError,
    APIBody
} from '../../utils/fetch-api';
import {
    ServiceList,
    ServiceError,
    ServiceDeleted,
} from './models';
export * from './models';

export const getListRequest = (baseURL: string, config: APIRequestInfo) => async <T>(path: string, page: number = 1, size: number = 10, filter: string = ''): Promise<APIResponseData<ServiceList<T>>|APIError<ServiceError>> => {
    const url = `${baseURL}${path}/?page=${page}&pageSize=${size}${filter ? `&search=${filter}` : ''}`;
    return fetchApi<undefined, ServiceList<T>, ServiceError>(url, {...config, method: 'GET'});
};
export const deleteRequest = (baseURL: string, config: APIRequestInfo) => async (path: string, entityIds: string|string[]): Promise<(APIResponseData<ServiceDeleted[]>|APIError<ServiceError>)> => {
    const entityArray = ((Array.isArray(entityIds)) ? entityIds : [entityIds]).filter(id => !!id);

    if(!entityIds || entityArray.length < 1) {
        return {
            errorCode: 500,
            errorMessage: '',
            error: {
                statusCode: 500,
                error: 'Missing id',
                message: 'id is an invalid id value or array'
            }
        };
    }
    const requests = [];
    for (const entityId of entityArray) {
        const url = `${baseURL}${path}/${entityId}`;
        requests.push(fetchApi<undefined, undefined, ServiceError>(url, {...config, method: 'DELETE'}));
    }
    const responses = await Promise.all(requests);
    const data: ServiceDeleted[] = responses.map((response, idx) => {
        const deletedResponse: ServiceDeleted = {
            id: entityArray[idx],
            state: 'DELETED'
        };
        if(isAPIError(response)) {
            deletedResponse.state = 'REJECTED';
            deletedResponse.error = response.error;
        }
        return deletedResponse;
    });
    return {
        status: 200,
        data
    };
};
export const createRequest = (baseURL: string, config: APIRequestInfo) => async <T extends APIBody>(path: string, body: Partial<T>): Promise<APIResponseData<T>|APIError<ServiceError>> => {
    const url = `${baseURL}${path}`;
    return fetchApi<Partial<T>, T, ServiceError>(url, {...config, method: 'POST', headers: {...config.headers, 'content-type': 'application/json'}, body});
};
export type Api = {
    getListRequest<T>(path: string, page?: number, size?: number, filter?: string): Promise<APIResponseData<ServiceList<T>>|APIError<ServiceError>>;
    deleteRequest(path: string, ids: string|string[]): Promise<(APIResponseData<ServiceDeleted[]>|APIError<ServiceError>)>;
    createRequest<T extends APIBody>(path: string, body: Partial<T>): Promise<APIResponseData<T>|APIError<ServiceError>>;
};
export const buildApiService = (server: string): Api => {
    const config: APIRequestInfo = {
        method: 'GET',
        headers: {}
    };
    return {
        getListRequest: getListRequest(server, config),
        deleteRequest: deleteRequest(server, config),
        createRequest: createRequest(server, config)
    };
};
