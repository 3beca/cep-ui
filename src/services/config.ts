export const BASE_URL = window._env_.BASE_URL;
export const EVENT_TYPE_URL = '/admin/event-types';