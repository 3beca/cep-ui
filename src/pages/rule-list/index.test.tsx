import * as React from 'react';
import {
    renderInsideApp,
    waitFor,
    setupNock,
    serverGetRuleList,
    generateRuleListWith,
    screen,
    act,
    within,
    waitForElementToBeRemoved,
    serverDeleteRule
} from '../../test-utils';
import userEvent from '@testing-library/user-event';
import RuleListPage from './index';
import { BASE_URL } from '../../services/config';
import { Rule, ServiceError, ServiceList } from '../../services/api';

export const selectCreateRule = async (ariaLabel: 'real time' | 'hopping' | 'sliding' | 'tumbling') => {
    const addButton = await screen.findByLabelText(/add rule/i);
    userEvent.click(addButton);
    await screen.findByLabelText(/title create rule/i);
    await screen.findByLabelText(/kind of rules description/i);
    const realTimeCard = await screen.findByLabelText(`create rule ${ariaLabel} card`);
    const selectButton = await screen.findByText(/^select$/i);
    userEvent.click(realTimeCard);
    userEvent.click(selectButton);
    await waitFor(() => expect(document.getElementById('create-rule-dialog')).toBe(null));
};

const initialPage = 1;
const initialPageSize = 10;
const filter = '';

test('RuleListPage should render loading component', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(initialPageSize, false, false));
    const { container } = renderInsideApp(<RuleListPage />);

    await screen.findByTestId(/loading-view-row/i);
    await screen.findByLabelText(/add rule/i);

    expect(container).toMatchSnapshot();

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
});

test('RuleListPage should render Empty List when no elements and snap', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(0, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findByTestId(/empty-view-row/i)).toBeInTheDocument();
    expect(await screen.findByTestId(/empty-view-row/i)).toHaveTextContent(/there are no rules created yet/i);
    await screen.findByLabelText(/add rule/i);
});

test('RuleListPage should render End List message when no more elements to load and snap', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(10, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(10);
    expect(await screen.findByTestId(/empty-view-row/i)).toHaveTextContent(/you reached the end of the list/i);
    await screen.findByLabelText(/add rule/i);
});

test('RuleListPage should render 20 cards and a create button and snapshot', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(initialPageSize, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/filters card rule/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/status card rule/i)).toHaveLength(initialPageSize);
    await screen.findByLabelText(/add rule/i);
});

test('RuleListPage should render a create rule dialog when click on add button and close when press close button', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(1, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(1);
    const addButton = await screen.findByLabelText(/add rule/i);

    userEvent.click(addButton);
    await screen.findByLabelText(/title create rule/i);
    await screen.findByLabelText(/kind of rules description/i);
    const closeButton = await screen.findByText(/^close$/i);

    userEvent.click(closeButton);
    await waitFor(() => expect(document.getElementById('create-rule-dialog')).toBe(null));
});

test('RuleListPage should render a create rule dialog when click on add button and close when press select button', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(initialPageSize, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
    await selectCreateRule('real time');
});

test('RuleListPage should render a searchbar  and find rules that contains rule', async () => {
    jest.useFakeTimers();
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(initialPageSize, false, false));
    renderInsideApp(<RuleListPage />);

    await screen.findByLabelText(/rule search bar/i);
    const input = (await screen.findByLabelText(/search input/i)) as HTMLInputElement;
    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
    await screen.findByLabelText(/add rule/i);

    const searchText = 'rule-name';
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, searchText, 200, generateRuleListWith(5, false, false));
    await userEvent.type(input, searchText);
    act(() => void jest.runOnlyPendingTimers());
    await screen.findByTestId(/loading-view-row/i);
    await waitFor(() => expect(screen.queryByTestId(/loading-view-row/i)).not.toBeInTheDocument());
    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(5);
    jest.useRealTimers();
});

test('RuleListPage should render Empty List whith filter when no elements and snap', async () => {
    jest.useFakeTimers();
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(0, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findByTestId(/empty-view-row/i)).toBeInTheDocument();
    expect(await screen.findByTestId(/empty-view-row/i)).toHaveTextContent(/there are no rules created yet/i);
    await screen.findByLabelText(/add rule/i);

    const searchText = 'rule-name';
    const input = (await screen.findByLabelText(/search input/i)) as HTMLInputElement;
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, searchText, 200, generateRuleListWith(0, false, false));
    await userEvent.type(input, searchText);
    act(() => void jest.runOnlyPendingTimers());
    await waitFor(() => expect(screen.queryByTestId(/loading-view-row/i)).not.toBeInTheDocument());
    expect(await screen.findByTestId(/empty-view-row/i)).toBeInTheDocument();
    expect(await screen.findByTestId(/empty-view-row/i)).toHaveTextContent(/There are no elements for "rule-name"/i);
    jest.useRealTimers();
});

test('RuleListPage should not render load button when there are NO MORE elements', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(10, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(10);
    expect(screen.queryByLabelText(/load more rules/i)).not.toBeInTheDocument();
    expect(await screen.findByTestId(/empty-view-row/i)).toHaveTextContent(/you reached the end of the list/i);
});

test('RuleListPage should render load button when there are more elements and load next set of elelments when click', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(initialPageSize, true, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
    const loadMore = await screen.findByLabelText(/load more rules/i);

    serverGetRuleList(setupNock(BASE_URL), initialPage + 1, initialPageSize, filter, 200, generateRuleListWith(5, false, true));
    userEvent.click(loadMore);
    expect(await screen.findByTestId(/loading-view-row/i)).toBeInTheDocument();
    await waitFor(() => expect(screen.queryByTestId(/loading-view-row/i)).not.toBeInTheDocument());
    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(15);
    expect(screen.queryByLabelText(/load more rules/i)).not.toBeInTheDocument();
    expect(await screen.findByTestId(/empty-view-row/i)).toHaveTextContent(/you reached the end of the list/i);
});

test('RuleListPage should render 10 RuleCards with header eventType, target, filters and status', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(initialPageSize, true, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/element card rule/i)).toHaveLength(10);
    expect(await screen.findAllByLabelText(/eventType name card rule/i)).toHaveLength(10);
    expect(await screen.findAllByLabelText(/target name card rule/i)).toHaveLength(10);
    expect(await screen.findAllByLabelText(/filters card rule/i)).toHaveLength(10);
    expect(await screen.findAllByLabelText(/status card rule/i)).toHaveLength(10);
});

test('RuleListPage should render a create rule dialog when click on add button and navigate to create realtime rule when select', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(1, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(1);
    await selectCreateRule('real time');
});

test('RuleListPage should render a create rule dialog when click on add button and navigate to create sliding rule when select', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(1, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(1);
    await selectCreateRule('sliding');
});

test('RuleListPage should render a create rule dialog when click on add button and navigate to create hopping rule when select', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(1, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(1);
    await selectCreateRule('hopping');
});

test('RuleListPage should render a create rule dialog when click on add button and navigate to create tumbling rule when select', async () => {
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, generateRuleListWith(1, false, false));
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(1);
    await selectCreateRule('tumbling');
});

test('RuleListPage should delete the first rule and update the list', async () => {
    const rules = generateRuleListWith(initialPageSize, true, false);
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, rules);
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/filters card rule/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/status card rule/i)).toHaveLength(initialPageSize);
    await screen.findByLabelText(/add rule/i);
    // Find first Rule Card
    const ruleCards = await screen.findAllByLabelText(/^element card rule$/i);
    const firstCard = within(ruleCards[0]);
    const ruleToDelete = rules.results[0];
    await screen.findByText(ruleToDelete.name);
    // Open acctions dialog
    userEvent.click(await firstCard.findByLabelText(/settings card rule/i));
    // Cacture actions dialog
    const dialogs = await screen.findAllByLabelText(/setting dialog card rule visible/i);
    expect(dialogs).toHaveLength(1);
    const actionsDialog = within(dialogs[0]);
    // Open delete action
    userEvent.click(await actionsDialog.findByLabelText(/setting dialog delete card rule/i));
    const deleteDialogs = await screen.findAllByLabelText(/delete dialog card rule/i);
    expect(deleteDialogs).toHaveLength(1);
    // Delete first Rule
    const newPageSize = initialPageSize - 1;
    serverDeleteRule(setupNock(BASE_URL), rules.results[0].id);
    const newRulesResult: ServiceList<Rule> = {
        ...rules,
        results: rules.results.slice(1)
    };
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, newRulesResult);
    userEvent.click(await screen.findByLabelText(/delete button/i));
    await waitForElementToBeRemoved(await screen.findByLabelText(/delete dialog card rule/i));
    await waitForElementToBeRemoved(await screen.findByTestId('loading-view-row'));
    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(newPageSize);
    expect(await screen.findAllByLabelText(/filters card rule/i)).toHaveLength(newPageSize);
    expect(await screen.findAllByLabelText(/status card rule/i)).toHaveLength(newPageSize);
    expect(screen.queryByText(ruleToDelete.name)).not.toBeInTheDocument();
});

test('RuleListPage should delete the first rule when load two pages and update the list', async () => {
    const rules = generateRuleListWith(initialPageSize, true, false, 'page1');
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, rules);
    const { unmount } = renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/filters card rule/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/status card rule/i)).toHaveLength(initialPageSize);
    await screen.findByLabelText(/add rule/i);

    // Load next page of rules
    const loadMore = await screen.findByLabelText(/load more rules/i);
    serverGetRuleList(setupNock(BASE_URL), initialPage + 1, initialPageSize, filter, 200, generateRuleListWith(5, false, true, 'page2'));
    userEvent.click(loadMore);
    expect(await screen.findByTestId(/loading-view-row/i)).toBeInTheDocument();
    await waitFor(() => expect(screen.queryByTestId(/loading-view-row/i)).not.toBeInTheDocument());
    const newPageSize = 15;
    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(newPageSize);
    expect(screen.queryByLabelText(/load more rules/i)).not.toBeInTheDocument();
    expect(await screen.findByTestId(/empty-view-row/i)).toHaveTextContent(/you reached the end of the list/i);

    // Find Rule Card in position 6
    const ruleCards = await screen.findAllByLabelText(/^element card rule$/i);
    const firstCard = within(ruleCards[6]);
    const ruleToDelete = rules.results[6];
    await screen.findByText(ruleToDelete.name);
    // Open acctions dialog
    userEvent.click(await firstCard.findByLabelText(/settings card rule/i));
    // Capture actions dialog
    const dialogs = await screen.findAllByLabelText(/setting dialog card rule visible/i);
    expect(dialogs).toHaveLength(1);
    const actionsDialog = within(dialogs[0]);
    // Open delete action
    userEvent.click(await actionsDialog.findByLabelText(/setting dialog delete card rule/i));
    const deleteDialogs = await screen.findAllByLabelText(/delete dialog card rule/i);
    expect(deleteDialogs).toHaveLength(1);
    // Delete Rule in position 6
    serverDeleteRule(setupNock(BASE_URL), ruleToDelete.id);
    userEvent.click(await screen.findByLabelText(/delete button/i));
    await waitForElementToBeRemoved(await screen.findByLabelText(/delete dialog card rule/i));
    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(newPageSize - 1);
    expect(await screen.findAllByLabelText(/filters card rule/i)).toHaveLength(newPageSize - 1);
    expect(await screen.findAllByLabelText(/status card rule/i)).toHaveLength(newPageSize - 1);
    expect(screen.queryByText(ruleToDelete.name)).not.toBeInTheDocument();
    unmount();
});

test('RuleListPage should fails when cannnot delete a Rule', async () => {
    const rules = generateRuleListWith(initialPageSize, true, false);
    serverGetRuleList(setupNock(BASE_URL), initialPage, initialPageSize, filter, 200, rules);
    renderInsideApp(<RuleListPage />);

    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/filters card rule/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/status card rule/i)).toHaveLength(initialPageSize);
    await screen.findByLabelText(/add rule/i);
    // Find first Rule Card
    const ruleCards = await screen.findAllByLabelText(/^element card rule$/i);
    const firstCard = within(ruleCards[0]);
    const ruleToDelete = rules.results[0];
    await screen.findByText(ruleToDelete.name);
    // Open acctions dialog
    userEvent.click(await firstCard.findByLabelText(/settings card rule/i));
    // Cacture actions dialog
    const dialogs = await screen.findAllByLabelText(/setting dialog card rule visible/i);
    expect(dialogs).toHaveLength(1);
    const actionsDialog = within(dialogs[0]);
    // Open delete action
    userEvent.click(await actionsDialog.findByLabelText(/setting dialog delete card rule/i));
    const deleteDialogs = await screen.findAllByLabelText(/delete dialog card rule/i);
    expect(deleteDialogs).toHaveLength(1);
    // Fails to delete first Rule
    const deleteError: ServiceError = {
        error: 'Invalid id of Rule',
        message: 'Cannot delete this rule, please go away...',
        statusCode: 400
    };
    serverDeleteRule(setupNock(BASE_URL), ruleToDelete.id, 400, deleteError);
    userEvent.click(await screen.findByLabelText(/delete button/i));
    // Show error message
    await screen.findByLabelText(/error message/i);
    // Close dialog
    userEvent.click(await screen.findByLabelText(/close button/));
    await waitForElementToBeRemoved(await screen.findByLabelText(/delete dialog card rule/i));
    expect(await screen.findAllByLabelText(/^element card rule$/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/filters card rule/i)).toHaveLength(initialPageSize);
    expect(await screen.findAllByLabelText(/status card rule/i)).toHaveLength(initialPageSize);
    expect(await screen.findByText(ruleToDelete.name)).toBeInTheDocument();
});
