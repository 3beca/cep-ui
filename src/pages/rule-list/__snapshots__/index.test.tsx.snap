// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleListPage should render Empty List whith filter when no elements and snap 1`] = `
<div>
  <div
    class="makeStyles-root-378"
  >
    <div
      aria-label="rule search bar"
      class="makeStyles-searchBar-379"
    >
      <div
        class="MuiPaper-root makeStyles-searchContainer-384 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiInputBase-root makeStyles-searchInput-385"
        >
          <input
            aria-label="search input"
            class="MuiInputBase-input"
            placeholder="Enter a rule name..."
            type="text"
            value="rule-name"
          />
        </div>
        <div
          aria-label="search icon"
          class="makeStyles-searchButton-386"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
      </div>
    </div>
    <button
      aria-label="add rule"
      class="MuiButtonBase-root MuiFab-root makeStyles-fabAddRule-381 MuiFab-primary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="makeStyles-gridCards-380"
    />
    <div
      class="makeStyles-emptyView-390"
      data-testid="empty-view-row"
    >
      <h4
        class="MuiTypography-root makeStyles-emptyViewText-391 MuiTypography-h4 MuiTypography-alignCenter"
      >
        There are no elements for "rule-name"
      </h4>
    </div>
  </div>
</div>
`;

exports[`RuleListPage should render loading component 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      aria-label="rule search bar"
      class="makeStyles-searchBar-2"
    >
      <div
        class="MuiPaper-root makeStyles-searchContainer-7 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiInputBase-root makeStyles-searchInput-8"
        >
          <input
            aria-label="search input"
            class="MuiInputBase-input"
            placeholder="Enter a rule name..."
            type="text"
            value=""
          />
        </div>
        <div
          aria-label="search icon"
          class="makeStyles-searchButton-9"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
      </div>
    </div>
    <button
      aria-label="add rule"
      class="MuiButtonBase-root MuiFab-root makeStyles-fabAddRule-4 MuiFab-primary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="makeStyles-gridCards-3"
    />
    <div
      class="makeStyles-loadingView-10"
      data-testid="loading-view-row"
    >
      <div
        class="MuiCircularProgress-root makeStyles-loadingSpinner-11 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
